src/libcore/alloc.rs

```diff
@@ -140,6 +140,37 @@ impl Layout {
         unsafe { Layout::from_size_align_unchecked(size, align) }
     }
 
+    /// Produces layout describing a record that could be used to
+    /// allocate backing structure for `T` (which could be a trait
+    /// or other unsized type like a slice).
+    ///
+    /// # Safety
+    ///
+    /// This function is only safe to call if one of the following conditions hold:
+    ///
+    /// - If `T` is `Sized`, this function is always safe to call.
+    /// - If the unsized tail of `T` is:
+    ///     - a [slice], then the length of the slice tail must be an intialized
+    ///       integer, and the size of the *entire value*
+    ///       (dynamic tail length + statically sized prefix) must fit in `isize`.
+    ///     - a [trait object], then the vtable part of the pointer must point
+    ///       to a valid vtable acquired by an unsizing coersion, and the size
+    ///       of the *entire value* (dynamic tail length + statically sized prefix)
+    ///       must fit in `isize`.
+    ///     - otherwise, it is conservatively not allowed to call this function.
```

Comment:
So the plan is to not allow this function on extern types? Or did you just not mention them because they are unstable?

If it is the latter, I feel we should mention them somehow -- maybe just saying that this is talking about an unstable language feature. But the question is inevitably going to come up.

Reply:
I believe it's "more proper" to disallow asking for extern types through this API, as they don't actually have a knowable size/align (from Rust's POV). However, depending on the final decision for calling the reference-based versions, the most consistent result may even change; perhaps we can use these functions to return an `Option` that's `None` for extern types.

IOW, I've not said that `extern type` is defined to give to this function because I honestly don't know what semantics are best to give them.

Reply:
It would be weird to have a semantics that's any different from `size_of_val`, wouldn't it? `extern type` has no metadata, just like sized types, after all.

Reply:
Of course, `size_of_val`'s semantics are [to panic in the face of `extern type` (most likely)](https://github.com/rust-lang/rust/issues/49708)

Reply:
Seems reasonable then to document that this will behave just like `size_of_val`?

Reply:
I have "but may panic or otherwise return the wrong value, as the extern type's layout is not known" currently. Given extern type is still unstable and the implementation of `[size|align]_of_val` don't yet match (what I _think_ is) consensus, I think this loose "safe but otherwise unspecified" is _probably_ enough to land these functions, and the wording can be re-addressed by the PR clarifying the actual intended behavior.

Reply:
I agree having such a vague wording makes sense. But people will also wonder "is there any difference to calling `size_of_val`", won't they?

Reply:
I added "This is the same behavior as [`fn`] on a reference to an extern type tail", so this should be adequate now.